lxd/images.go

```diff
@@ -1115,93 +1115,223 @@ func autoUpdateImagesTask(d *Daemon) (task.Func, task.Schedule) {
 }
 
 func autoUpdateImages(ctx context.Context, d *Daemon) error {
-	projectNames := []string{}
+	imageMap := make(map[string][]db.Image)
+
 	err := d.cluster.Transaction(func(tx *db.ClusterTx) error {
-		projects, err := tx.GetProjects(db.ProjectFilter{})
+		var err error
+
+		images, err := tx.GetImages(db.ImageFilter{AutoUpdate: true})
 		if err != nil {
 			return err
 		}
 
-		for _, project := range projects {
-			projectNames = append(projectNames, project.Name)
+		for _, image := range images {
+			imageMap[image.Fingerprint] = append(imageMap[image.Fingerprint], image)
 		}
 
 		return nil
 	})
 	if err != nil {
-		return errors.Wrap(err, "Unable to retrieve project names")
+		return errors.Wrap(err, "Unable to retrieve image fingerprints")
 	}
 
-	for _, project := range projectNames {
-		err := autoUpdateImagesInProject(ctx, d, project)
+	for fingerprint, images := range imageMap {
+		nodes, err := d.cluster.GetNodesWithImageAndAutoUpdate(fingerprint, true)
 		if err != nil {
-			return errors.Wrapf(err, "Unable to update images for project %s", project)
+			logger.Error("Error getting cluster members for image auto-update", log.Ctx{"fingerprint": fingerprint, "err": err})
+			continue
+		}
+
+		if len(nodes) == 0 {
+			continue
+		}
+
+		if len(nodes) > 1 {
+			var nodeIDs []int64
+
+			for _, node := range nodes {
+				err := d.cluster.Transaction(func(tx *db.ClusterTx) error {
+					var err error
+
+					nodeInfo, err := tx.GetNodeByAddress(node)
+					if err != nil {
+						return err
+					}
+
+					nodeIDs = append(nodeIDs, nodeInfo.ID)
+
+					return nil
+				})
+				if err != nil {
+					logger.Error("Unable to retrieve cluster member information for image update", log.Ctx{"err": err})
+					continue
```

Comment:
@monstermunchkin I think we need to skip the whole image update if this happens, not just the retrieving the node info, because otherwise we may end up incorrectly thinking there are only 1 nodes and then handling the update on more than 1 nodes at once.